src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1111,280 @@ private static string RegularExpressionFromFileSpec
+            && !HasTags(fixedDirectoryPart)
+            && !HasTags(wildcardDirectoryPart)
+            && !HasTags(filenamePart);
+
+        private static bool HasDotDot(string str)
+        {
+            for (int i = 0; i < str.Length - 1; i++)
             {
-                isLegalFileSpec = false;
-                return String.Empty;
+                if (str[i] == '.' && str[i + 1] == '.')
+                {
+                    return true;
+                }
             }
+            return false;
+        }
 
-            /* 
-             * Trailing dots in file names have to be treated specially.
-             * We want:
-             * 
-             *     *. to match foo
-             * 
-             * but 'foo' doesn't have a trailing '.' so we need to handle this while still being careful 
-             * not to match 'foo.txt'
-             */
-            if (filenamePart.EndsWith(".", StringComparison.Ordinal))
+        private static bool HasMisplacedRecursiveOperator(string str)
+        {
+            for (int i = 0; i < str.Length - 1; i++)
             {
-                filenamePart = filenamePart.Replace("*", "<:anythingbutdot:>");
-                filenamePart = filenamePart.Replace("?", "<:anysinglecharacterbutdot:>");
-                filenamePart = filenamePart.Substring(0, filenamePart.Length - 1);
-            }
+                bool isRecursiveOperator = str[i] == '*' && str[i + 1] == '*';
 
-            /*
-             * Now, build up the starting filespec but put tags in to identify where the fixedDirectory,
-             * wildcardDirectory and filenamePart are. Also tag the beginning of the line and the end of
-             * the line, so that we can identify patterns by whether they're on one end or the other.
-             */
-            StringBuilder matchFileExpression = new StringBuilder();
-            matchFileExpression.Append("<:bol:>");
-            matchFileExpression.Append("<:fixeddir:>").Append(fixedDirectoryPart).Append("<:endfixeddir:>");
-            matchFileExpression.Append("<:wildcarddir:>").Append(wildcardDirectoryPart).Append("<:endwildcarddir:>");
-            matchFileExpression.Append("<:filename:>").Append(filenamePart).Append("<:endfilename:>");
-            matchFileExpression.Append("<:eol:>");
+                // Check boundaries for cases such as **\foo\ and *.cs**
+                bool isSurroundedBySlashes = (i == 0 || FileUtilities.IsAnySlash(str[i - 1]))
+                                             && i < str.Length - 2 && FileUtilities.IsAnySlash(str[i + 2]);
 
-            /*
-             *  Call out our special matching characters.
-             */
-            foreach (var separator in directorySeparatorStrings)
+                if (isRecursiveOperator && !isSurroundedBySlashes)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private static bool HasTags(string str)
```

Comment:
Assuming this check isn't needed anymore but left it in since it'll break some tests expecting an invalid itemspec

Reply:
It's a behavior change since <: and :> would be now allowed in an the filespec, I think that seems reasonable

Reply:
Yeah, I guess they're illegal now because they're illegal win32 FS characters so we'd fall back anyway? But when you remove most of the checks things get super confusing?